go/chat/boxer_test.go

```diff
@@ -1488,3 +1499,71 @@ func TestRemarshalBoxed(t *testing.T) {
 
 	require.Equal(t, boxed1.ClientHeader.OutboxID, boxed2.ClientHeader.OutboxID, "obids should have same value")
 }
+
+func randomTeamEK() keybase1.TeamEk {
+	randBytes, err := libkb.RandBytes(32)
+	if err != nil {
+		panic(err)
+	}
+	seed := libkb.MakeByte32(randBytes)
+	dhKey := (*ephemeral.TeamEKSeed)(&seed).DeriveDHKey()
+	teamEK := keybase1.TeamEk{
+		Seed: seed,
+		Metadata: keybase1.TeamEkMetadata{
+			Kid: dhKey.GetKID(),
+		},
+	}
+	return teamEK
+}
+
+func TestExplodingMessageUnbox(t *testing.T) {
+	key := cryptKey(t)
+	ephemeralKey := randomTeamEK()
+	text := "hello exploding"
+	tc, boxer := setupChatTest(t, "exploding")
+	defer tc.Cleanup()
+	// We need a user for unboxing to work.
+	u, err := kbtest.CreateAndSignupFakeUser("unbox", tc.G)
+	if err != nil {
+		t.Fatal(err)
+	}
+	msg := textMsgWithSender(t, text, gregor1.UID(u.User.GetUID().ToBytes()), chat1.MessageBoxedVersion_V3)
+
+	// Set the ephemeral metadata, to indicate that the messages is exploding.
+	msg.ClientHeader.EphemeralMetadata = &chat1.MsgEphemeralMetadata{
+		Lifetime: 99999,
+	}
+
+	// Box it! Note that we pass in the ephemeral/exploding key, and also set
+	// V3 explicitly.
+	boxed, err := boxer.box(context.TODO(), msg, key, &ephemeralKey, getSigningKeyPairForTest(t, tc, u), chat1.MessageBoxedVersion_V3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.Equal(t, chat1.MessageBoxedVersion_V3, boxed.Version)
+	if len(boxed.BodyCiphertext.E) == 0 {
+		t.Error("after boxMessage, BodyCipherText.E is empty")
+	}
+
+	// We need to give it a server header for unboxing...
+	boxed.ServerHeader = &chat1.MessageServerHeader{
+		Ctime: gregor1.ToTime(time.Now()),
+	}
+
+	// Unbox it!!!
+	unboxed, err := boxer.unbox(context.TODO(), *boxed, chat1.ConversationMembersType_TEAM, key, &ephemeralKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	body := unboxed.MessageBody
+	if typ, _ := body.MessageType(); typ != chat1.MessageType_TEXT {
+		t.Errorf("body type: %d, expected %d", typ, chat1.MessageType_TEXT)
+	}
+	if body.Text().Body != text {
+		t.Errorf("body text: %q, expected %q", body.Text().Body, text)
+	}
+	require.Nil(t, unboxed.SenderDeviceRevokedAt, "message should not be from revoked device")
+	require.NotNil(t, unboxed.BodyHash)
+	require.True(t, unboxed.IsExploding())
+	require.NotNil(t, unboxed.EphemeralMetadata())
```

Comment:
i'd sanity check equality to the metadata you boxed here

Reply:
Added an equality lifetime check. (The generation field is set by boxing and won't be equal.)